### PR TITLE
[jvm-packages] support distributed synchronization of customized evaluation metrics

### PR DESCRIPTION
**Pull request by CodingCat**
- [x] basic JNI framework ( c++ <-> java )
- [x] support binary classification
- [x] support regression
- [x] support multi classes
- [x] support ranking
- [x] unit test checking numerical correctness of synchronized metrics
- [x] document 

this PR implements the functionality to synchronize custom metrics in distributed training setting. Before this PR, only the built-in metrics are synchronized resulting that users cannot leverage functionalities like early-stopping with custom metrics (since different workers may have different decision about whether to stop).

This PR is to 

* implement distributed synchronization of custom metrics 
* avoid users who implement the metrics to explicitly call rabit operation

closes https://github.com/dmlc/xgboost/issues/3595 https://github.com/dmlc/xgboost/issues/3813

### CODE REVIEW

**Comment by thvasilo** [comment]
Hello @CodingCat , could you write a couple of things on the purpose of this so I can potentially help out with reviewing and testing?

**Comment by CodingCat** [comment]
@thvasilo thanks for offering the help on testing the feature, I will have a more comprehensive description of the PR after I finalize the code part

**Comment by CodingCat** [comment]
@thvasilo thanks for offering the help on testing the feature, I will have a more comprehensive description of the PR after I finalize the code part

**Comment by CodingCat** [comment]
@trivialfis @RAMitchell @hcho3 

please help to review this, as I moved several classes in the original .cu files to headers to be shared by other files

**Comment by trivialfis** [comment]
Had a brief look.  This is gonna take some time.

File: include/xgboost/learner.h [hunk 1]
```diff
@@ -186,15 +186,16 @@ class Learner : public rabit::Serializable {
    */
   virtual const std::map<std::string, std::string>& GetConfigurationArguments() const = 0;
 
+  /*! \brief The evaluation metrics used to evaluate the model. */
+  std::vector<std::unique_ptr<Metric> > metrics_;
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply include/xgboost/learner.h]: sorry for the late reply, just back from the vacation, because I referred metrics_ in https://github.com/dmlc/xgboost/pull/4280/files#diff-3288207540b433242fe8da1390c68d3fR1164

File: include/xgboost/metric/metric.h [hunk 2]
```diff
@@ -1,21 +1,20 @@
...
 #include <dmlc/registry.h>
 #include <vector>
 #include <string>
 #include <functional>
 #include <utility>
 
-#include "./data.h"
-#include "./base.h"
-#include "../../src/common/host_device_vector.h"
+#include "../data.h"
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply include/xgboost/metric/metric.h]: done

File: include/xgboost/c_api.h [hunk 3]
```diff
@@ -20,6 +20,8 @@
 // XGBoost C API will include APIs in Rabit C API
 #include <rabit/c_api.h>
 
+#include <string>
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply include/xgboost/c_api.h]: done

**Comment by CodingCat** [comment]
> My concern with this PR is that it fuses the internal C++ implementation of metrics to the external Java API. If we merge this we effectively lose the ability to refactor or change the way metrics are implemented.
> 
> While this provides some generality for users it still doesn't give us the ability to implement metrics like AUC via Java/Scala.
> 
> Given how complicated it seems to expose this feature and the fact that it is not a complete solution, I wonder if users should be implementing these functions in C++?

@RAMitchell , thanks for the review

actually we have make an regulation on the signature of methods in metrics implementations, e.g. https://github.com/dmlc/xgboost/blob/2e052e74b627659b97ce741be5b64581479303c3/src/metric/elementwise_metric.cu#L106, 

as long as we want to expose the distributed metrics syncing to users without requiring them to call rabit directly (which is not an elegant solution as we are exposing an XGBoost dependency to the user), they need to follow that signature no matter which language they want to 

regarding implementing something like AUC, the users can do that with existing interface like https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/EvalTrait.scala#L38 (they just cannot use it in a distributed setting for purposes like early stopping)

the question is only on do we want to `expose the distributed metrics syncing to without requiring them to call rabit directly `?

**Comment by CodingCat** [comment]
> @CodingCat Let's don't merge this for now.
> 
> * Is the feature jvm only or it's universal to other bindings (I have never done this before)?  If it's jvm only then we can discuss how to make it more generic before moving so many code.
> * Is it necessary to move the whole implementation into header?  We can create a suitable interface.
> * Other stuffs in comments.

for now, only JVM is having this functionality as the strategy is to make cross-language calls via language-specific APIs....

IIRC, other languages didn't provide any base class for users to extend and implement custom metrics, instead, they require the user to implement the whole function , like https://github.com/dmlc/xgboost/blob/82dca3c108502b6912f2135c3120a24a3ec55045/python-package/xgboost/core.py#L1147. Then the users needs to care about everything from logic to distributed sync

if we want to extend to other languages, I think the pattern here is a good references though cannot be applied directly in implementation (because every language has its own cross-language function call APIs)

**Comment by CodingCat** [comment]
can we take further consideration on this PR?

**Comment by CodingCat** [comment]
@trivialfis @hcho3 @RAMitchell I got failed compilation in Jenkins 

it says 

```

[2019-04-19T03:53:37.599Z] In file included from /workspace/src/metric/multiclass_metric.cc:8:0:

[2019-04-19T03:53:37.599Z] /workspace/include/xgboost/metric/multiclass_metric.h:17:59: fatal error: thrust/execution_policy.h: No such file or directory

[2019-04-19T03:53:37.599Z] compilation terminated.

```

do you have any insights about what is happening?

**Comment by trivialfis** [comment]
@CodingCat Let me try fixing it in another PR.

**Comment by CodingCat** [comment]
nvm, I think I messed up the switch for conditional compilation

**Comment by trivialfis** [comment]
@CodingCat The thrust header paths are managed by `nvcc`, which will pass the right arguments and device code to gcc.  For a `cu` file this all happens automatically.  But for a `cc` file, you need guards like `__CUDACC__` or `XGBOOST_USE_CUDA` to avoid letting gcc process CUDA related code directly.

**Comment by CodingCat** [comment]
> @CodingCat The thrust header paths are managed by `nvcc`, which will pass the right arguments and device code to gcc. For a `cu` file this all happens automatically. But for a `cc` file, you need guards like `__CUDACC__` or `XGBOOST_USE_CUDA` to avoid letting gcc process CUDA related code directly.

got it! thanks!

**Comment by CodingCat** [comment]
ping for the further review @trivialfis @RAMitchell @hcho3

**Comment by RAMitchell** [comment]
Can we explore the alternative a little more? Could we expose a very basic AllReduce method via the xgboost API that connects on the back end to the rabit library linked to the xgboost shared library?

**Comment by CodingCat** [comment]
we actually have exposed something like https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/native/xgboost4j.h#L327-L328

do you mean we are going to document how to implement a customized metrics for distributed training with this API (instead of masking this with the implementation here)?

**Comment by RAMitchell** [comment]
> do you mean we are going to document how to implement a customized metrics for distributed training with this API (instead of masking the call to rabit api with the implementation here)?

I guess I would like to know what it look like and how complicated it would be for users to call rabit themselves in their jvm implementation of the metric. If it's not overly difficult and distributed metrics can be achieved this way it may be preferable to this current implementation for me.

**Comment by CodingCat** [comment]
@RAMitchell I was trying to write an example to concretely show how it is complicated for a user to write a distributed metrics by their own, but discarded just because of the complexity, 

given the current interface https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/IEvaluation.java#L40

1. in the first step, users need to get <b>predictions</b>, <b>labels</b> and <b>weights</b>

2. then they need to implement something like https://github.com/dmlc/xgboost/blob/master/src/metric/elementwise_metric.cu#L36-L57

3. after that, they need to call Rabit API to synchronize residual and weights, like https://github.com/dmlc/xgboost/blob/master/src/metric/elementwise_metric.cu#L36-L57

4. another headache is that we need to either provide a specific API tailored for syncing float array or the user has to use our current rabit api but deal with transform their local evaluation results got from step 1, 2 to a nio.directBuffer in java and pass to our API...

I would still prefer to mask all these complexities to the users, since we have implemented everything in C++, it's not worth repeating it in Java....and I am actually unclear about the downside of this approach except it's a big diff most of which are just from moving files?

**Comment by RAMitchell** [comment]
IMO the solution would be make it easy for the users to get predictions, labels and weights on the JVM side and a function simplifying the rabit reduction. The reason why I am taking this angle as it requires no changes to the C API and leaves us much more flexible in terms of how we internally implement metrics in C++ code. I don't think this solution is more complicated than what you have.

I realise you have done a lot of work already but I want to make sure this is correct as I think it will have long-term implications on my work and others. Lets look for other opinions, in particular I'm curious what @tqchen thinks.

**Comment by CodingCat** [comment]
One thing I would add is that we should think about how we treat metrics system in xgb, is it like a helper function which is fine under any changed or we are going to treat it as a submodule sharing by other modules

If we want to make it a formal module, we need to sign the contract to regulate the changes

The contract here is, no matter how you change it, you should allow others to tell 

1. What’s the output given the prediction and label of a single row

2. What’s the output given the sum of errors and weights

re on more opinions

**Comment by CodingCat** [comment]
I will get back on this after 0.9 release

and until so far, I have no idea about 

(1) why we require every user to write the same chunk of code like https://github.com/dmlc/xgboost/blob/master/src/metric/elementwise_metric.cu#L36-L57  if they want to use custom evaluation metrics in distributed setting

(2) why we require XGBoost user knows what Rabit is? (Do we also require our user knows what NCCL is and how to use it?) 

(3) why we are so afraid of stablizing any API even we are seriously considering 1.0 release

(4) what is the concrete plan of metrics system refactoring (as concrete as the solution which have been here for two months) which can be broken by exposing any API?

<b>I don't think we are on the right track to move things forward </b>...such disaster happened in other communities as well and essentially hurt users and create many private folks of the project. One of the most significant example might be the nested column pruning feature in SparkSQL, that's a feature proposed 2 years ago, but was delayed time and time again due to the reason like "hmmmm, we need to refactor/re-design/whatever A,B,C, there <b>might</b> be some conflict" (turns out no in most of cases), even that's a feature waited by many users....

and eventually they only merged a (very very) partial version of that feature, and pissed-off users have created some private folks of Spark containing that feature and many other innovations, for example, FB/Uber/Apple

**Comment by hcho3** [comment]
@CodingCat One thing I like about this pull request is `IEvalMultiClassesDistributed`. I've seen quite a few users tripping over writing customized objectives for multi-class classification, e.g. #4288. It turns out that writing customized objective for multi-class classification is hard to do correctly, even in non-distributed setting. I'd love to see first-class support for this use case.

**Comment by hcho3** [comment]
@CodingCat One little useful detail: AUC / AUCPR are actually list-wise ranking metric. So `IEvalRankListDistributed` should be able to accommodate AUC / AUCPR. (In the case of binary classification, set the number of query groups to 1.)

File: doc/jvm/xgboost4j_spark_tutorial.rst [hunk 4]
```diff
@@ -205,6 +205,18 @@ Training with Evaluation Sets
...
 
+Training with Custom Evaluation Metrics
+----------------
+With XGBoost4j (including XGBoost4J-Spark), users are able to implement their own custom evaluation metrics and synchronize the values in the distributed training setting. To implement a custom evaluation metric, users should implement the interface ``ml.dmlc.xgboost4j.java.IEvalElementWiseDistributed`` (for binary classification and regression), ``ml.dmlc.xgboost4j.java.IEvalMultiClassesDistributed`` (for multi classification) and ``ml.dmlc.xgboost4j.java.IEvalRankListDistributed`` (for ranking).
+
+* ``ml.dmlc.xgboost4j.java.IEvalElementWiseDistributed``: users are supposed to implement ``float evalRow(float label, float pred);`` which calculates the metric for a single sample given the prediction and label, as well as ``float getFinal(float errorSum, float weightSum);`` which performs the final transformation over the sum of error and weights of samples.
```
Comment (hcho3) [inline doc/jvm/xgboost4j_spark_tutorial.rst]: For completeness, we need to have weights included in element-wise metric:
```
float evalRow(float label, float pred, float weight)
```

We may also want to define an interface to let the user define a custom reduce op.

File: doc/jvm/xgboost4j_spark_tutorial.rst [hunk 5]
```diff
@@ -205,6 +205,18 @@ Training with Evaluation Sets
...
+
+* ``ml.dmlc.xgboost4j.java.IEvalMultiClassesDistributed``: the methods to be implemented by the users are similar to ``ml.dmlc.xgboost4j.java.IEvalElementWiseDistributed`` except that the single row metric calculating method is ``float evalRow(int label, float pred, int numClasses);``
```
Comment (hcho3) [inline doc/jvm/xgboost4j_spark_tutorial.rst]: Should be ``float evalRow(int label, float pred, float weight, int numClasses);``

File: doc/jvm/xgboost4j_spark_tutorial.rst [hunk 6]
```diff
@@ -205,6 +205,18 @@ Training with Evaluation Sets
...
+
+* ``ml.dmlc.xgboost4j.java.IEvalRankListDistributed``: users are to implement ``float evalMetric(float[] preds, int[] labels);`` which gives the predictions and labels for instances in the same group;
```
Comment (hcho3) [inline doc/jvm/xgboost4j_spark_tutorial.rst]: Should be `float evalMetric(float[] preds, int[] labels, float group_weight);`

**Comment by trivialfis** [comment]
@CodingCat I will see if it's possible to "unmask" the sum op.
